cats/CI_api_interface.py

```diff
@@ -0,0 +1,84 @@
+from dataclasses import dataclass
+from datetime import datetime, timezone
+
+@dataclass(order=True)
+class CarbonIntensityEstimate:
+    """Represents an average carbon intensity for a timeperiod.
+    Use order=True in order to enable comparison of class
+    instance based on the sort_index attribute. See
+    https://peps.python.org/pep-0557
+
+    value: needs to be in gCO2e
+
+    """
+    value: float # value needs to be first for the sort function to work
+    start: datetime
+    end: datetime
+
+    def __post_init__(self):
+        self.sort_index = self.value # TODO check whether that's useful
```

Comment:
It allows to compare two instances of the class. Generally useful if you ask me, but particularly useful in the context of the `WindowedForecast` implementation to be able to write `min(WindowedForecast(data, size))`, see #36

Reply:
Yes sorting is important, still key to the current implementation where the list of `CarbonIntensityEstimate` is being sorted. However, i don't think that this line is needed to enable the sort, rather what counts is to put `value` in first position. As described in [this StackOverflow post](https://stackoverflow.com/questions/68840767/what-is-the-purpose-of-sort-index-of-a-dataclass-in-python), `self.sort_index` seems to be only needed is the sorting index is not provided when building the dataclass instance (not the case here). Or I may have missed something?

Here is a quick piece of code that tests that:

```python
@dataclass(order=True)
    class valueFirst:
        value: float
        start: datetime
        end: datetime
        def __post_init__(self):
            # self.sort_index = self.value
            self.timedelta = self.end - self.start

foo = valueFirst(start=datetime(2023,5,20,15,30), end=datetime(2023,5,21,15,40), value=18)
bar = valueFirst(start=datetime(2023, 5, 22, 15, 30), end=datetime(2023, 5, 23, 15, 40), value=12)

@dataclass(order=True)
class startFirst_with_index:
    start: datetime
    end: datetime
    value: float
    def __post_init__(self):
        self.sort_index = self.value
        self.timedelta = self.end - self.start

fooo = startFirst_with_index(start=datetime(2023, 5, 20, 15, 30), end=datetime(2023, 5, 21, 15, 40), value=18)
baar = startFirst_with_index(start=datetime(2023, 5, 22, 15, 30), end=datetime(2023, 5, 23, 15, 40), value=12)
```
`foo < bar` returns `False`, as expected (foo.value > bar.value)
But `fooo < baar` returns `True`, because `fooo.start < baar.start` (and the `sort_index` doesn't seem to have an impact)

Reply:
With `order=True` comparison between dataclasses is a tuple comparison (i.e `('b', 2) > ('a', 4)` is `True`. So your first example works because `value` is the first fields and that's what you want the comparison to be based on. Not the case in your second example and yet it works as expected thanks to setting `self.sort_index`.

Reply:
But the second example doesn't work as expected, `fooo.value=18` while `baar.value=12`, so the expected behaviour would be `fooo < baar` to be False, which is not the case here.

Reply:
Sorry - I read too fast. I didn't realise `sort_index` is not a special attribute and as you observed, does not have any special meaning on its own. Had to go back reading a bit more on dataclasses  -- it appears to be a trick to impose ordering according to an specific class attribute other than the first one (you probably know this already if you read that SO post linked above).

```python
@dataclass(order=True)
class CarbonIntensityEstimate:
    sort_index: int = field(init = False, repr = False)
    value: float
    start: datetime
    end: datetime

    def __post_init__(self):
        self.sort_index = self.value
```

Or can just make sure that `value` is listed first. But specifying the sort_index makes it explicitly how instances of the class are orderered? Your call.

Reply:
I'm thinking of adding a comment to highlight what is being used, especially as `value` is not the only parameter used (in case of equal value, it goes to the next one, and we want to use `start` to start the job as quickly as possible), so the sequence of parameter is important. And that lightens the code. If something like below sounds sensible?

```python
@dataclass(order=True)
class CarbonIntensityEstimate:
    # instances can compared and ordered based on the order of the parameters below.
    # Here, functions such as <, >, min, max will use value first, and in case of equal values, use start
    value: float
    start: datetime
    end: datetime
```